// decision/lib/decisionChain.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _chainstyle = require("chainstyle");

var _chainstyle2 = _interopRequireDefault(_chainstyle);

// TODO filter

var Decision = (0, _chainstyle2["default"])({
    when: {
        checkType: ["string"]
    },
    then: {
        checkType: ["string | function"]
    },
    filter: {
        checkType: ["string | function"]
    },
    data: {
        checkType: ["function"]
    },
    named: {
        checkType: ["string"]
    }
}, {}, {
    chainRegular: /^when\.(filter\.)?(data\.)?then(\.named)?$/
});

exports["default"] = function (marmoset, alleyway) {
    var decisions = [];
    return {
        createFactory: function createFactory() {
            return function () {
                var decision = new Decision();
                decisions.push(decision);
                return decision;
            };
        },
        getDecisions: function getDecisions() {
            return decisions;
        }
    };
};

module.exports = exports["default"];
