// marmoset/lib/eventTraceline/traceRuleAdapter.js
/**
 * createTraceEvent((m) => {});
 * 
 * e1.e2      m().happen(ctx1, "e1").inCase((e)=>{}).trigger(ctx2, "e2", data)
 * e1.e2.e3   m().happen(ctx2, "e1.e2").inCase((e)=>{}).trigger(ctx3, "e3", data)
 *
 * lastEventType
 * case
 * nextEvent
 */

"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _chainstyle = require("chainstyle");

var _chainstyle2 = _interopRequireDefault(_chainstyle);

var TraceAdapter = (0, _chainstyle2["default"])({
    happen: {
        checkType: ["any", "string"]
    },
    inCase: {
        checkType: ["function"]
    },
    trigger: {
        checkType: ["any", "string"]
    },
    data: {
        checkType: ["function"]
    }
}, {
    getRule: {
        method: function method() {
            return this.end(function (queueInfo) {
                var map = queueInfo.getMap();
                return {
                    lastEventType: {
                        ctx: map["happen"].args[0],
                        type: map["happen"].args[1]
                    },
                    "case": map["inCase"].args[0],
                    nextEvent: {
                        ctx: map["trigger"].args[0],
                        type: map["trigger"].args[1],
                        data: map["trigger"].args[2],
                        dataRule: map["data"] && map["data"].args[0]
                    }
                };
            });
        }
    }
}, {
    chainRegular: /^happen\.inCase\.trigger(\.data)?$/
});

exports["default"] = function () {
    var insts = [];
    return {
        createRuleFactory: function createRuleFactory() {
            return function () {
                var inst = new TraceAdapter();
                for (var name in inst) {}
                insts.push(inst);
                return inst;
            };
        },
        getRules: function getRules() {
            var rules = [];
            for (var i = 0; i < insts.length; i++) {
                var inst = insts[i];
                rules.push(inst.getRule());
            }
            return rules;
        }
    };
};

module.exports = exports["default"];
