// marmoset/lib/closeEventOrder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

exports["default"] = function (timelineModel) {
    var createCloseOrderEvent = function createCloseOrderEvent(ctx, type) {
        var sentences = type.split(";");
        var ctxs = [];
        for (var i = 0; i < sentences.length; i++) {
            ctxs.push(ctx);
        }
        createCloseGroupEvent(ctxs, type);
    };

    var createCloseGroupEvent = function createCloseGroupEvent(ctxs, type) {
        var types = type.split(";");
        while (types.length > 1) {
            var last = types.pop();
            var rest = types.join(";");
            addCloseEvent(ctxs, last, rest);
        }
    };

    var addCloseEvent = function addCloseEvent(ctxs, last, rest) {
        var ctxLast = ctxs.pop();
        var ctxRest = ctxs[ctxs.length - 1];
        timelineModel.builders.createEventByTimeline(ctxLast, rest + ";" + last, function (start, m, dataMerge) {
            start.match([ctxLast], [last]);

            m().interest([ctxLast], [last]).interest([ctxRest], [rest]).lookBackAt(1).should(function (e) {
                if (!e) return 0;
                if (e.type == rest) return 1;
                return 0;
            });

            dataMerge.merge(function (cur) {
                return cur.data;
            });
        });
    };
    return {
        onModels: [createCloseOrderEvent],
        onGroupModels: [createCloseGroupEvent],
        specialMetas: [";"]
    };
};

module.exports = exports["default"];
