// marmoset/lib/eventTimeline/adapter/timelineInterface.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _chainstyle = require("chainstyle");

var _chainstyle2 = _interopRequireDefault(_chainstyle);

/**
 * SETMRule
 */

var SETMRule = (0, _chainstyle2["default"])({
    interest: {
        checkType: ["falsy | array", "string | array | regExp"]
    },
    should: {
        checkType: ["function"]
    },
    lookBackAt: {
        checkType: ["number"]
    },
    lookForwardAt: {
        checkType: ["number"]
    }
}, {}, {
    chainRegular: /^((interest\.)+lookBackAt\.should)|((interest\.)+lookForwardAt\.should)$/
});

/**
 * Data merge
 */
var DataMergeRule = function DataMergeRule() {};

DataMergeRule.prototype = {
    constructor: DataMergeRule,
    merge: function merge(mergeHandler) {
        this.mergeHandler = mergeHandler;
    }
};

/**
 * TriggerEventRule
 *
 * match
 */

var TriggerEventRule = function TriggerEventRule() {};

var TriggerEventRule = (0, _chainstyle2["default"])({
    match: {
        checkType: ["falsy | array", "string | array | regExp"]
    }
}, {
    chainRegular: /^match$/
});

exports["default"] = {
    SETMRule: SETMRule,
    TriggerEventRule: TriggerEventRule,
    DataMergeRule: DataMergeRule
};
module.exports = exports["default"];
